packages/devtools_app/lib/src/timeline/timeline_flame_chart.dart

```diff
@@ -353,9 +354,47 @@ class FullTimelineFlameChartCanvas extends FlameChartCanvas<FullTimelineData> {
 
             // Horizontal guidelines connecting each child to the vertical
             // guideline above.
-            for (var child in event.children) {
+            for (int i = 0; i < event.children.length; i++) {
+              double horizontalGuidelineStartX;
+
+              final child = event.children[i];
               final childNode = chartNodesByEvent[child];
-              final horizontalGuidelineStartX = verticalGuidelineX;
+
+              AsyncTimelineEvent previousChild;
+              if (i > 0) previousChild = event.children[i - 1];
+
+              // If [child] does not overlap [previousChild], draw the vertical
+              // guideline for [child] near the left edge of [child] instead of
+              // on the left edge of [event].
+              if (previousChild != null && !previousChild.deepOverlaps(child)) {
+                final previousChildNode = chartNodesByEvent[previousChild];
+
+                // If [child] started after [event] ended, use the right edge of
+                // event's [node] as the x coordinate for the guideline.
+                // Otherwise, take the minimum of
+                // [subsequentChildGuidelineOffset] and half the distance
+                // between [previousChild] and [child].
+                double newVerticalGuidelineX;
+                if (event.time.end < child.time.start) {
+                  newVerticalGuidelineX = node.rect.right;
+                } else {
+                  newVerticalGuidelineX = childNode.rect.left -
+                      math.min(
+                        subsequentChildGuidelineOffset,
+                        (childNode.rect.left - previousChildNode.rect.right) /
+                            2,
+                      );
+                }
+                final newVerticalGuidelineEndY = childNode.rect.centerLeft.dy;
+                verticalGuidelines.add(VerticalLineSegment(
+                  Offset(newVerticalGuidelineX, verticalGuidelineStartY),
+                  Offset(newVerticalGuidelineX, newVerticalGuidelineEndY),
+                ));
+
+                horizontalGuidelineStartX = newVerticalGuidelineX;
+              } else {
+                horizontalGuidelineStartX = verticalGuidelineX;
```

Comment:
nit: maybe instead initialize this before the for loop and have the for loop start at `i=1`
that would let you remove the i >0 and the != previousChild != null checks.